crates/typst/src/eval/call.rs

```diff
@@ -368,6 +261,111 @@ pub(crate) fn call_closure(
+///
+/// This follows the normal function call order: we evaluate the callee before the
+/// arguments.
+///
+/// Prioritize associated functions on the value's type (i.e., methods) over its fields.
+/// A function call on a field is only allowed for functions, types, modules (because
+/// they are scopes), and symbols (because they have modifiers).
+///
+/// For dictionaries, it is not allowed because it would be ambiguous (prioritizing
+/// associated functions would make an addition of a new associated function a breaking
+/// change and prioritizing fields would break associated functions for certain
+/// dictionaries).
+fn eval_method_call(
+    target_expr: ast::Expr,
+    method: Ident,
+    args: ast::Args,
+    span: Span,
+    vm: &mut Vm,
+) -> SourceResult<MethodCall> {
+    // Evaluate the method's target and overall arguments.
+    let (target, mut args) = if is_mutating_method(&method) {
+        // If `method` looks like a mutating method, we evaluate the arguments first,
+        // because `target_expr.access(vm)` mutably borrows the `vm`, so that we can't
+        // evaluate the arguments after it.
+        let args = args.eval(vm)?.spanned(span);
+        // However, this difference from the normal call order is not observable because
+        // expressions like, i.e. `(1, arr.len(), 2, 3).push(arr.pop())`, evaluate the
+        // target to a temporary which we disallow mutation on (returning an error).
+        // Theoretically this could be observed if a method matching `is_mutating_method`
+        // was added to some type in the future and we didn't update this function.
+        match target_expr.access(vm)? {
+            // Only arrays and dictionaries have mutable methods.
+            target @ (Value::Array(_) | Value::Dict(_)) => {
+                let value = call_method_mut(target, &method, args, span);
+                let point = || Tracepoint::Call(Some(method.get().clone()));
+                return Ok(MethodCall::Special(value.trace(vm.world(), point, span)?));
+            }
+            target => (target.clone(), args),
+        }
+    } else {
+        let target = target_expr.eval(vm)?;
+        let args = args.eval(vm)?.spanned(span);
+        (target, args)
+    };
+
+    if let Value::Plugin(plugin) = &target {
+        // Call plugin methods by converting args to bytes.
+        let bytes = args.all::<Bytes>()?;
+        args.finish()?;
+        let value = plugin.call(&method, bytes).at(span)?.into_value();
+        Ok(MethodCall::Special(value))
+    } else if let Some(callee) = target.ty().scope().get(&method) {
+        args.insert(0, target_expr.span(), target);
+        Ok(MethodCall::Normal(callee.clone(), args))
+    } else if matches!(
+        target,
+        Value::Symbol(_) | Value::Func(_) | Value::Type(_) | Value::Module(_)
+    ) {
+        // Certain value types may have their own ways to access method fields.
+        // e.g. `$arrow.r(v)$`, `func.with(fill: red)`
```

Comment:
`func.with` is a wrong example here, I think. This branch covers things like `table.cell[..]`. `func.with` is handled by `target.ty().scope().get(&method)`.